File: CPP/SieveOfEratosthenes.cpp
#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

void sieveOfEratosthenes(int start, int end){
    bool *primes = new bool[end + 1];
    for(int i = 0; i <= end; i++){
        primes[i] = true;
    }
    primes[0] = false;
    primes[1] = false;
    for(int i = 2; i <= sqrt(end); i++){
        if (primes[i]){
            int multiple = i * 2;
            while (multiple <= end){
                primes[multiple] = false;
                multiple += i;
            }
        }
    }
    for(int i = start; i <= end; i++){
        if(primes[i]){
            cout << i << " is a prime number!" << endl;
        }
    }
    delete [] primes;
}

vector<int> sieveOfEratosthenesList(int start, int end){
    vector<int> primeList;    
    bool *primes = new bool[end + 1];
    for(int i = 0; i <= end; i++){
        primes[i] = true;
    }
    primes[0] = false;
    primes[1] = false;
    for(int i = 2; i <= sqrt(end); i++){
        if (primes[i]){
            int multiple = i * 2;
            while (multiple <= end){
                primes[multiple] = false;
                multiple += i;
            }
        }
    }
    for(int i = start; i <= end; i++){
        if(primes[i]){
            primeList.push_back(i);
        }
    }
    delete [] primes;
    return primeList;
}

int main() {
    sieveOfEratosthenes(20, 353);

    vector<int> sieveResult;
    sieveResult = sieveOfEratosthenesList(20, 353);
    for (int i = 0; i < sieveResult.size(); i++) {
        cout << sieveResult[i] << ", ";
    }
}

File: Python/SieveOfEratosthenes.py
#O(nlogn)
def sieve_of_eratosthenes(start: int, end: int) -> None:
    primes = (end + 1) * [True]
    primes[0] = primes[1] = False
    for i in range(2, int(end**0.5) + 1):
        if primes[i]:
            multiple = i * 2
            while multiple <= end:
                primes[multiple] = False
                multiple += i
    for i in range(start, end + 1):
        if primes[i]:
            print(str(i) + " is a prime number!")

sieve_of_eratosthenes(20, 353)

#O(nlogn)
def sieve_of_eratosthenes_list(start: int, end: int) -> list:
    prime_list = []
    primes = (end + 1) * [True]
    primes[0] = primes[1] = False
    for i in range(2, int(end**0.5) + 1):
        if primes[i]:
            multiple = i * 2
            while multiple <= end:
                primes[multiple] = False
                multiple += i
    for i in range(start, end + 1):
        if primes[i]:
            prime_list.append(i)
    return prime_list

print(sieve_of_eratosthenes_list(20, 353))